tests/functional/base.py

```diff
@@ -0,0 +1,316 @@
+import os
+import configparser
+import tempfile
+import json
+from tempest.lib.cli import base
+from tempest.lib.cli import output_parser
+from tempest.lib.common.utils import data_utils
+
+class EsiFunctionalBase(base.ClientTestBase):
+    @classmethod
+    def setUpClass(cls):
+        super(EsiFunctionalBase, cls).setUpClass()
+        cls.clients = {}
+        cls.config = {}
+        cls.client_info = {}
+        cls._cleanups = []
+
+        cls._init_config(cls)
+        cls._init_client(cls, 'admin')
+        cls._init_roles(cls)
+
+    @classmethod
+    def tearDownClass(cls):
+        cls._cleanups.reverse()
+        for cleanup in cls._cleanups:
+            cleanup[0](*cleanup[1:])
+
+    def _get_clients(self):
+        #NOTE: ClientTestBase's constructor requires this to be implemented, but
+        #to initialize our clients, we need ClientTestBase's constructor to have
+        #been called. We return nothing and leave the job up to _init_clients().
+        return {}
+
+    def _init_config(self):
+        venv_name = os.environ.get('OS_VENV_NAME', default='functional')
+        self.config['cli_dir'] = os.path.join(os.path.abspath('.'), 
+                '.tox/%s/bin' % venv_name)
+
+        # allow custom configuration to be passed in via env var
+        cfg_file_path = os.environ.get('OS_ESI_CFG_PATH',
+                '/etc/esi-leap/esi-leap.conf')
+        cfg_parser = configparser.ConfigParser()
+        if not cfg_parser.read(cfg_file_path):
+            self.fail('Could not open config file %s for reading' %
+                    cfg_file_path)
+            return -1
+
+        # attempts to read authentication credentials in this order:
+        # 1) [keystone] section of config file
+        # 2) [keystone_authtoken] section of config file
+        # 3) from the environment variables (e.g. OS_PASSWORD, etc)
+        def auth_get(val):
+            for sect in 'keystone', 'keystone_authtoken':
+                if cfg_parser.has_section(sect):
+                    try:
+                        return cfg_parser.get(sect, val)
+                    except:
+                        pass
+            x = os.environ.get('OS_%s' % val.upper())
+            if x != None:
```

Comment:
it is preferred to use the `is` or `is not` when comparing to `None` instead of `==` or `!=`. 
```python
if x is not None:
    return x
```


https://www.python.org/dev/peps/pep-0008/#programming-recommendations

Reply:
More bad habits from other programming languages slipping into my Python code...

Reply:
no worries! I would not exactly call them bad habits, just that python has a different style I guess. :smile: